mypy/messages.py

```diff
@@ -472,7 +517,16 @@ def format_distinctly(self, type1: Type, type2: Type, bare: bool = False) -> Tup
     # get some information as arguments, and they build an error message based
     # on them.
 
-    def has_no_attr(self, original_type: Type, typ: Type, member: str, context: Context) -> Type:
+    @tracked()
+    def no_return_exepected(self, context: Context) -> None:
+        self.fail('No return value expected', context)
```

Comment:
Converting all static messages to this format may cause lots of code churn, but in principle I am fine with this.

Another possible option is to have a global dictionary (or JSON file) mapping error message to error code:
```python
message_to_id = {
    NO_RETURN_VALUE_EXPECTED: 'unexpected_return',
    MUST_HAVE_NONE_RETURN_TYPE: 'unexpected_return',
    MISSING_RETURN_STATEMENT: 'missing_return',
    ...
}
```

Reply:
> Another possible option is to have a global dictionary (or JSON file) mapping error message to error code

I think that could be a good compromise, but it doesn't work for messages that require formatting so we'd have to be very clear about this.  A developer might expect this to work with the `message_to_id` dict:

```python
self.fail(messages.MUST_HAVE_NONE_RETURN_TYPE.format(fdef.name()), item)
```

We can protect against this in a few ways:
- clear comments near `message_to_id` documenting that this is only for static messages
- code that checks `message_to_id` keys for formatting operations `{}`.  
- add a variation of `MessageBuilder.fail`, or a flag, that raises an exception if the message is not in `message_to_id` (I'm assuming here that we're not going to get to 100% message id coverage in the near term)

Reply:
I just pushed a commit inspired your `message_to_id` idea, which addresses most of my concerns about type safety.  I used an `Enum`, where the names are ids and the values are the messages.   This is used for messages that are static or require only basic formatting (though I haven transitioned all of the code yet).  If you like it, I'll finish it up.

Reply:
@ilevkivskyi  Sorry to bug you.  Can I get a sign off on the idea of using `Enum`?  If that sounds like a good idea to you I think I can take this PR to completion.

Reply:
NP, I will think more about this tomorrow and let you know, I was really busy at work last two days.

Reply:
Eh, it looks like we can't use enums for two reasons:
* Mypyc compiler doesn't support custom metaclasses yet (`Enum` has a complex metaclass), there are plans to support it, but not in near few months.
* More importantly, plugin users should be able to produce custom error messages, which is hard with enums.

I thought more about possible strategies, and it looks like having a global dictionary for all messages is probably inevitable. I like your idea of changing the signature of `MessageBuilder.fail()` to accept "base" message and a list of format parameters. If possible, I would even put `format` before `context` and make it a required argument (caller would pass an empty tuple if the base message is just a constant), but probably we can do this later. Here are some things that I would propose:
* Focus on core logic rather on completeness in this PR
* Add a global dictionary `message_registry: Dict[str, str]` in `messages.py`, add only 3-5 keys to the dictionary in this PR.
* Only add `@register()` to 3-5 methods
* The `fail()` method will first try to get the error code from `message_registry` using base message as a key, if the base message is not there, try getting the active message (btw, why this needs to be a stack, not just a single value?), if no luck, use `'uncategorized_error'` (all user plugin messages will be here).
* In few subsequent PRs gradually add `@register()` to more methods, and more keys to `message_registry`.
* When we will get a bit more experience we might want to migrate messages either to `@register()` or to `message_registry`. But we will probably have both because both have pros and cons. For example static or super simple errors a better for latter, while there are some error messages constructed in a tricky manner, for those the former is better.

What do you think?

Reply:
Everything you've written makes sense.  The main problem is that I can't get over the notion that the `message_registry` dictionary is going to be awfully redundant and tedious to maintain, since it's mostly going to be the message constant followed by a lowercase string of the message constant's name (depending on how broad vs granular we are with grouping) :

```python
message_registry = {
    NO_RETURN_VALUE_EXPECTED: 'no_return_value_expected',
    MISSING_RETURN_STATEMENT: 'missing_return_statement',
    INVALID_IMPLICIT_RETURN: 'invalid_implicit_return',
    INCOMPATIBLE_RETURN_VALUE_TYPE: 'incompatible_return_type',
    RETURN_VALUE_EXPECTED: 'return_value_expected',
    NO_RETURN_EXPECTED: 'no_return_expected',
    INVALID_EXCEPTION: 'invalid_exception',
    INVALID_EXCEPTION_TYPE: 'invalid_exception',
    RETURN_IN_ASYNC_GENERATOR: 'return_in_async_generator',
}
```

I also don't like that the message and the id are stored in different locations.  I think making the constants into basic data classes would be an improvement:

```python
NO_RETURN_VALUE_EXPECTED = Msg('No return value expected', 'no_return_value_expected')  # type: Final
MISSING_RETURN_STATEMENT = Msg('Missing return statement', 'missing_return_statement')  # type: Final
```

But I'd really love to design a system where the constant doubles as the id, *unless* it's overridden by a group definition.  If enums won't work, the only other way I can think of to do this is with a descriptor.  Should I just give up on this idea?

Reply:
> But I'd really love to design a system where the constant doubles as the id

Why not move all the constants into a dedicated module, and iterate over the module namespace and calculate the id based on the attribute name? This sort of metaprogramming should be okay with mypyc. Something like this:

```py
...
FOO_BAR = Msg('Foo bar')
MSG_B = Msg('Another message')

for name, val in list(globals().items()):
    if isinstance(val, Msg) and val.id is None:
        val.id = name.lower()
```

Reply:
> Gradual transition to the new scheme for existing error messages (I strongly prefer two-three smaller PRs, than one large)

How about I start with a new PR that moves all of the existing constants into `mypy.message_registry` with no other changes?  That seems to be one part of the design we've all come to agreement on.

Reply:
TBH, I don't see how this is important _now_. Anyway, we agreed that @JukkaL will follow on this and related PRs.